# **AdaBoost**

ここでは，ブースティングの応用アルゴリズムである AdaBoost (adaptive boosting) を解説します．ブースティングは弱学習器を複数組み合わせることで強力なモデルを作り出す手法ですが，その中でも AdaBoost は，シンプルで効果的なアルゴリズムとして広く知られています．

## **AdaBoost とは何か**

AdaBoost は，1996年に Y. Freund 氏と R. Schapire 氏によって提案された機械学習アルゴリズムです．ブースティングの考え方に基づき，複数の弱学習器を組み合わせることで，高精度な予測を行う強力な分類器や回帰モデルを構築することを目的としています．

従来の機械学習手法では，1つのモデルに依存することが一般的でした．しかし，特に弱学習器に当たる単純なモデル（例えば，浅い決定木や線形モデル）は，単独では限られた性能しか発揮できませんでした．そこで，**AdaBoost は弱学習器を段階的に強化し，全体として高い精度のモデルを作ることを可能にする手法**として注目を集めました．さらに，AdaBoost は分類問題だけでなく回帰問題にも適用可能であり，その柔軟性と効果の高さから多くの実用的な応用が見られます．

AdaBoost の回帰のアルゴリズムはいろいろな種類がありますが，以下では，各反復でサンプルの重みを更新し，誤差の大きいサンプルの重みを増加させる AdaBoost.R2 を扱うことにします．

## **AdaBoost による回帰のアルゴリズム**

AdaBoost による回帰のアルゴリズムの基本的なアイデアは，各段で予測誤差の大きいデータに重みを増やし，次ステップの弱学習器がその誤差に対してより正確に予測できるようにする点です．

以下，トレーニングデータは $N$ 個あり，各データ点を $(\boldsymbol{x}_{i},\ y_{i})\ (i = 1,2,\dots,N)$ とします．ここで，$y_{i}$ は特徴量 $\boldsymbol{x}_{i}$ に対するターゲットのラベルを表します．

<br>

#### **■Step 1. 初期重みの設定**

全てのトレーニングデータに対して等しい重みを付与します．トレーニングデータが $N$ 個ある場合，各データ $(\boldsymbol{x}_{i},\ y_{i})$ には次の式で表される重み（初期重み）$w_{i}^{(1)}$ が付与されます．

$$w_{i}^{(1)} = \frac{1}{N}$$

この式は，最初は全てのデータが同等に扱われることを意味しています．この先は，予測誤差の大きいデータ点の重みが増加し，次の学習ステップでそのデータ点がより重要視される仕組みです．

<br>

#### **■Step 2. 弱学習器の構築**

次に，**重み付けされたデータを用いて第** $m$ **段目の弱学習器** ${\widehat{f}}_{m}(\boldsymbol{x}_{i})$ **を構築**します．弱学習器は完全に正確である必要はなく，あくまでベースラインとなる予測器を構築します．

<br>

#### **■Step 3. 重み付き誤差の計算**

学習した弱学習器を用いて，各データ点 $(\boldsymbol{x}_{i},\ y_{i})$ に対して第 $m$ 段目の弱学習器の予測値 ${\widehat{f}}_{m}(\boldsymbol{x}_{i})$ を求め，各予測値の絶対誤差 (MAE) ${\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right)$ を考えます．

$${\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right) = |y_{i} - {\widehat{f}}_{m}(\boldsymbol{x}_{i})|$$

そのうえで，全データ点についての最大誤差 $D_{m} = \max_{i}{{\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right)}$ を使って正規化した損失関数として，次のようなものが用いられます．

- 線形損失：$\displaystyle L_{m}\left( \boldsymbol{x}_{i} \right) = \frac{{\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right)}{D_{m}} = \frac{|y_{i} - {\widehat{f}}_{m}(\boldsymbol{x}_{i})|}{D_{m}}$

- 二乗損失：$\displaystyle L_{m}\left( \boldsymbol{x}_{i} \right) = \left( \frac{{\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right)}{D_{m}} \right)^{2} = \left( \frac{|y_{i} - {\widehat{f}}_{m}(\boldsymbol{x}_{i})|}{D_{m}} \right)^{2}$

- 指数損失：$\displaystyle L_{m}\left( \boldsymbol{x}_{i} \right) = 1 - \exp\left( - \frac{{\widetilde{L}}_{m}\left( \boldsymbol{x}_{i} \right)}{D_{m}} \right) = 1 - \exp\left( - \frac{|y_{i} - {\widehat{f}}_{m}(\boldsymbol{x}_{i})|}{D_{m}} \right)$

以下では，線形損失（正規化した予測誤差）を用いることにします．この場合，$0 \leq L_{m}\left( \boldsymbol{x}_{i} \right) \leq 1$ となります．さらに，この損失関数をもとにした**重み付き誤差** $e_{m}$ を次の式で求めます．

$$e_{m} = \frac{\sum_{i = 1}^{N}\mspace{2mu}\mspace{2mu} w_{i}^{(m)}L_{m}\left( \boldsymbol{x}_{i} \right)}{\sum_{i = 1}^{N}\mspace{2mu}\mspace{2mu} w_{i}^{(m)}}$$

$e_{m}$ は，第$m$段目の弱学習器 ${\widehat{f}}_{m}$ の性能を示し，値が小さいほど学習器が全体としてより正確な予測を行っていることを意味します．また，$w_{i}^{(m)}$ は第 $m$ 段目の重みです．

<br>

#### **■Step 4. 重みの更新係数の計算**

**Step 3**で求めた $e_{m}$ を使って，**重みの更新係数** $\beta_{m}$ を計算します．

$$\beta_{m} = \frac{e_{m}}{1 - e_{m}}$$

$\beta_{m}$ は弱学習器の性能に応じて重みを更新するために使われます．$0 \leq \beta_{m}$ の範囲の値をとり，弱学習器の性能が良いほど $\beta_{m}$ は小さい値となります．

<br>

#### **■Step 5. データ点の重みの更新**

予測誤差が大きかったデータ点は，次の段でより重視されるように重みが更新されます．具体的には，次のように行います．

$$w_{i}^{(m + 1)} = w_{i}^{(m)} \cdot {\beta_{m}}^{1 - L_{m}\left( \boldsymbol{x}_{i} \right)}$$

ここで，$0 \leq L_{m}\left( \boldsymbol{x}_{i} \right) \leq 1，0 \leq \beta_{m}$ なので，$L_{m}\left( \boldsymbol{x}_{i} \right)$ が大きくなる（$L_{m}\left( \boldsymbol{x}_{i} \right) \rightarrow 1$ となる）と，

$$w_{i}^{(m + 1)} \fallingdotseq w_{i}^{(m)} \cdot {\beta_{m}}^{0} = w_{i}^{(m)}$$

となり，重みはほとんど変わりません．一方，$L_{m}\left( \boldsymbol{x}_{i} \right)$ が小さくなる（$L_{m}\left( \boldsymbol{x}_{i} \right) \rightarrow 0$ となる）と，

$$w_{i}^{(m + 1)} \fallingdotseq w_{i}^{(m)} \cdot {\beta_{m}}^{1} = w_{i}^{(m)} \cdot \beta_{m}$$

となり，重みは相対的に大きく減少することになります．

<br>

#### **■Step 6. 重みの正規化**

全てのデータ点の重みが合計で$1$になるように，重みを正規化します．

$$w_{i}^{(m + 1)} \leftarrow \frac{w_{i}^{(m + 1)}}{\sum_{j = 1}^{N}\mspace{2mu}\mspace{2mu} w_{j}^{(m + 1)}}$$

これにより，次の弱学習器が学習する際に，データ全体で重みのバランスが保たれます．

<br>

#### **■Step 7. 最終予測の計算**

第 $m$ 段目の弱学習器の予測値 $\widehat{f}_m\left( \boldsymbol{x}_{i} \right)$ を集計し，次のように最終的な予測値 ${\widehat{f}}_\mathrm{final}\left( \boldsymbol{x}_{i} \right)$ を決定します．

$${\widehat{f}}_\mathrm{final}\left( \boldsymbol{x}_{i} \right) = \inf\left\{ y \in Y:\sum_{m:{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) \leq y}^{}\mspace{2mu}\mspace{2mu}\log\left( \frac{1}{\beta_{m}} \right) \geq \frac{1}{2}\sum_{m}^{}\mspace{2mu}\mspace{2mu}\log\left( \frac{1}{\beta_{m}} \right) \right\}$$

ここで，$Y$ は全弱学習器の予測値の集合です．上式について詳しく説明すると，次のようになります．

1.  $e_{m}$ が小さい（予測精度が高い）ほど $\beta_{m}$ は小さくなるので，$\log(1/\beta_{m})$ は「その弱学習器の予測をどれだけ信用するか」を表す重みとして機能し，これを各弱学習器の**信頼度**と呼びます．

2.  $(1/2)\sum_{m}^{}\mspace{2mu}\mspace{2mu} \log(1/\beta_{m})$ は，各弱学習器の信頼度の総和の半分です．

3.  $\sum_{m:{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) \leq y}^{}\mspace{2mu}\mspace{2mu} \log(1/\beta_{m})$ は，予測値 ${\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right)$ が閾値 $y$ 以下であるような弱学習器の信頼度の総和です．

4.  $\sum_{m:{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) \leq y}^{}\mspace{2mu}\mspace{2mu} \log(1/\beta_{m})$ が $(1/2)\sum_{m}^{}\mspace{2mu}\mspace{2mu} \log(1/\beta_{m})$ 以上になる最小の $y$ の値を${\widehat{f}}_\mathrm{final}\left( \boldsymbol{x}_{i} \right)$ とします．

つまり，「ある閾値 $y$ 以下の学習器の信頼度の合計が，全ての学習器の信頼度の合計の50%以上になる最小の $y$」を最終的な予測値 ${\widehat{f}}_\mathrm{final}\left( \boldsymbol{x}_{i} \right)$ とします．この計算方法は**加重中央値**の計算であり，外れ値の影響を抑えつつ，全体的な予測精度を向上させる特徴があります．

## **AdaBoost による分類のアルゴリズム**

AdaBoost による分類のアルゴリズムの基本的なアイデアは，各段において誤分類したデータ点に対する重みを増加させ，次の弱学習器がその誤分類をより正確に予測できるようにする点にあります．これにより，各学習器が得意とするデータの特徴を活かしながら，最終的に強力な分類器を構築できます．

以下，トレーニングデータは $N$ 個あり，各データ点を $(\boldsymbol{x}_{i},\ y_{i})\ (i = 1,2,\dots,N)$ とします．ここで，特徴量 $\boldsymbol{x}_{i}$ に対するターゲットのラベル $y_{i}$ の値は$1$または$- 1$をとるものとします．

<br>

#### **■Step 1. 初期重みの設定**

全てのトレーニングデータに対して等しい重みを付与します．各データ $(\boldsymbol{x}_{i},\ y_{i})$ には次の式で表される重み（初期重み）$w_{i}^{(1)}$ が付与されます．

$$w_{i}^{(1)} = \frac{1}{N}$$

この式は，最初は全てのデータが同等に扱われることを意味しています．この先は，予測誤差の大きいデータ点の重みが増加し，次の学習ステップでそのデータ点がより重要視される仕組みです．

<br>

#### **■Step 2. 弱学習器の構築**

次に，**重み付けされたデータを用いて第** $m$ **段目の弱学習器** ${\widehat{f}}_{m}(\boldsymbol{x}_{i})$ **を構築**します．弱学習器は完全に正確である必要はなく，あくまでベースラインとなる予測器を構築します．

弱学習器 ${\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right)$ は単純なモデルであり，例えば決定木の深さを$1$に制限したモデル（スタンプ），または線形分類器が多く使用されます．弱学習器の出力は，各データ点 $\boldsymbol{x}_{\boldsymbol{i}}$ に対して$1$もしくは$- 1$を予測するものとします．

この段階では，弱学習器は完全な正解を出す必要はなく，誤分類率が50%を下回る性能であれば十分とされます．各学習ステップでは，前のステップで誤分類されたデータ点の重みが強調されているため，次の学習器はそれらのデータ点をより正確に分類することが期待されます．

<br>

#### **■Step 3. 重み付き誤分類率の計算**

弱学習器 ${\widehat{f}}_{m}$ が誤分類したデータの重みの割合を**重み付き誤分類率** $\varepsilon_{m}$ といい，次の式で計算されます．

$$\varepsilon_{m} = \frac{\sum_{i = 1}^{N}\mspace{2mu}\mspace{2mu} w_{i}^{(m)}\boldsymbol{1}\{ y_{i} \neq {\widehat{f}}_{m}(\boldsymbol{x}_{i})\}}{\sum_{i = 1}^{N}\mspace{2mu}\mspace{2mu} w_{i}^{(m)}}$$

ここで，$w_{i}^{(m)}$ は第 $m$ 段においてデータ点 $(\boldsymbol{x}_{i},\ y_{i})$ に付与された重み，$\boldsymbol{1}\{\cdot\}$ は指示関数です．

$\varepsilon_{m}$ が小さいほどモデルは優れているといえます．一方，$\varepsilon_{m} > 0.5$ の場合は，その学習器の予測精度はコイントスのようなランダムな分類よりも悪いことを意味します．この場合，学習器の符号を反転することを検討し，${\widehat{f}}_{m}$ の出力を $- {\widehat{f}}_{m}$ にすることで，性能の向上が可能となる場合があります．

<br>

#### **■Step 4. 弱学習器の重要度の計算**

**弱学習器の重要度**（第 $m$ 段目の弱学習器が最終予測にどれだけの影響を与えるかを示すパラメータ）$\alpha_{m}$ を，**Step 3**で求めた弱学習器の重み付き誤分類率 $\varepsilon_{m}$ を用いて，次の式で計算します．

$$\alpha_{m} = \frac{1}{2}\log\frac{1 - \varepsilon_{m}}{\varepsilon_{m}}$$

$\varepsilon_{m} \leq 0.5$ ならば

$$\frac{1 - \varepsilon_{m}}{\varepsilon_{m}} = \frac{1}{\varepsilon_{m}} - 1 \geq 1$$

なので，$\alpha_{m} \geq 0$ となります．このとき，弱学習器の予測はランダム予測より優れており，最終予測に加算することでモデルの精度向上に寄与します．なお，ランダム予測とは，入力データによらずクラスをランダムに割り当てる予測を指します．

一方，$\varepsilon_{m} > 0.5$ のとき（ランダム予測より悪い場合）は $\alpha_{m} < 0$ となり，そのまま加算すると精度に悪影響を及ぼすため，**Step 3**で述べた符号反転（${\widehat{f}}_{m}$の出力を$-{\widehat{f}}_{m}$ にする）が必要です．

<br>

#### **■Step 5. データ点の重みの更新**

**Step 4**で得られた $\alpha_{m}$ を用いて，次の段の弱学習器が学習する際の重み $w_{i}^{(m + 1)}$ を更新します．重みの更新式は次の通りです．

$$
\begin{equation}w_{i}^{(m + 1)} = w_{i}^{(m)} \cdot \exp\left( - \alpha_{m}y_{i}{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) \right) \tag{1}\end{equation}$$

誤分類されたデータ $\boldsymbol{x}_{i}$ に対しては，$y_{i}{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) = - 1$ となるので，式$(1)$より，

$$w_{i}^{(m + 1)} = w_{i}^{(m)} \cdot \exp\left( \alpha_{m} \right)$$

となります．また，**Step 1**と式$(1)$より，明らかに $w_{i}^{(m)} > 0$ です．

**Step 4**で述べたように，$\varepsilon_{m} \leq 0.5$ の場合，$\alpha_{m} \geq 0$ なので，$\exp\left( \alpha_{m} \right) \geq 1$ です．この場合は，弱学習器の重要度 $\alpha_{m}$ が大きいほど重みが増加し，次の段の弱学習器がそのデータに注目して学習することが期待されます．

一方，正しく分類されたデータ $\boldsymbol{x}_{i}$ に対しては，$y_{i}{\widehat{f}}_{m}\left( \boldsymbol{x}_{i} \right) = 1$ となるので，式$(1)$より，

$$w_{i}^{(m + 1)} = w_{i}^{(m)} \cdot \exp\left( - \alpha_{m} \right)$$

となります．また，$w_{i}^{(m)} > 0$ であり，さらに $\varepsilon_{m} \leq 0.5$ の場合は $\alpha_{m} \geq 0$ なので，$\exp\left( - \alpha_{m} \right) \leq 1$ です．したがって，このとき重みは減少するまたは変化しないことになります．

<br>

#### **■Step 6. 重みの正規化**

全てのデータ点の重みが合計で$1$になるように，次の式で正規化します．

$$w_{i}^{(m + 1)} \leftarrow \frac{w_{i}^{(m + 1)}}{\sum_{j = 1}^{N}\mspace{2mu}\mspace{2mu} w_{j}^{(m + 1)}}$$

これにより，次の弱学習器が学習する際に，データ全体で重みのバランスが保たれます．

<br>

#### **■Step 7. 最終的な分類**

最終的な予測は，全ての弱学習器の予測結果を加重平均し，それに基づいて決定されます．具体的には次の式で表されます．

$$\widehat{y}\left( \boldsymbol{x} \right) = \text{sign}\left( \sum_{m = 1}^{M}\mspace{2mu}\mspace{2mu}\alpha_{m}{\widehat{f}}_{m}\left( \boldsymbol{x} \right) \right) \tag{2}$$

ここで，sign関数は，引数として与えられた値が正か負かに応じて出力を決定する関数であり，次のように定義されます．

$$\text{sign}(z) = \left\{ \begin{matrix}
1 & (z > 0) \\
 - 1 & (z \leq 0)
\end{matrix} \right.\ $$

式$(2)$より，弱学習器の予測結果の加重和 $\sum_{m = 1}^{M}\mspace{2mu}\mspace{2mu}\alpha_{m}{\widehat{f}}_{m}(\boldsymbol{x})$ が正であれば「$1$」（正クラス），負であれば「$- 1$」（負クラス）が予測されます．

式$(2)$によって，各弱学習器の予測結果 ${\widehat{f}}_{m}(\boldsymbol{x})$ が重要度 $\alpha_{m}$ に基づいて重み付けされ，最終的な予測が行われます．性能の良い弱学習器ほど強く影響し，性能の低い学習器の影響は相対的に小さくなります．

## **AdaBoostの特徴と問題点**

AdaBoost の最大の特徴は，逐次的にデータの重みを適応的に更新することで，誤分類を段階的に減少させることです．また，過学習のリスクが比較的低いという利点もあります．これは，各弱学習器が単純なモデルであるため，個々の学習器がデータに過度に適合しにくいことによるものです．そのため，AdaBoost は幅広い機械学習タスクに適用可能であり，画像認識，テキスト分類，音声認識などの分野で成功を収めています．

一方で，AdaBoost にはいくつかの問題点もあります．特に，データにノイズや外れ値が多く含まれている場合，その影響を過大に受けやすいことがあります．誤分類されたデータに重みを増加させるため，ノイズや外れ値に対しても過度に重みが付与され，最終的なモデルの性能が低下する可能性があります．この問題を克服するために，ノイズに対してロバストな改良版の AdaBoost アルゴリズムも提案されています．

## **PythonでAdaBoost**

Python で AdaBoost を実装するには，scikit-learn ライブラリを使用するのが手軽です．ここでは，AdaBoost による分類モデルを作成します．

「Titanic Dataset」の乗船客の分類を scikit-learn の AdaBoost で実装してみましょう．

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# scikit-learnのAdaBoostと決定木をインポート
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

# 1. データの読み込みと前処理
print("1. データの読み込みと前処理...")

# データをロード
data = sns.load_dataset('titanic')

# 使用する特徴量とターゲット
features = ['pclass', 'sex', 'age', 'sibsp', 'parch', 'fare']
target = 'survived'

# 簡単な前処理
# 欠損値(age)を中央値で埋める
data['age'] = data['age'].fillna(data['age'].median())

# カテゴリ変数(Sex)を数値(0, 1)に変換
data['sex'] = data['sex'].map({'male': 0, 'female': 1})

# 'embark_town' などの欠損値がある行を削除
data = data.dropna(subset=['embark_town'])

X = data[features].values
# sklearn版では y は {0, 1} のままでOK
y = data[target].values

# 訓練データとテストデータに分割
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(f"訓練データ: {X_train.shape[0]}件, テストデータ: {X_test.shape[0]}件")

# ----------------------------------------------------

print("\n2. AdaBoostモデルの学習 (scikit-learn版)...")

# 弱い学習器（深さ1の決定木）を定義
base_learner = DecisionTreeClassifier(max_depth=1)

# AdaBoostモデルを初期化
# n_estimators: 弱い学習器の数
# estimator: 弱い学習器のモデル
# random_state: 再現性のためのシード値
ada_model = AdaBoostClassifier(
    estimator=base_learner,
    n_estimators=50,
    random_state=42
)

# モデルの学習 (fit)
ada_model.fit(X_train, y_train)

print("学習完了．")

# ----------------------------------------------------

print("\n3. モデルの評価...")

# 訓練データの予測と精度
y_pred_train = ada_model.predict(X_train)
train_accuracy = accuracy_score(y_train, y_pred_train)
print(f"訓練データの正解率 (Accuracy): {train_accuracy:.4f}")

# テストデータの予測と精度
y_pred_test = ada_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred_test)
print(f"テストデータの正解率 (Accuracy): {test_accuracy:.4f}")

# 混同行列の表示
print("\nテストデータの混同行列:")
cm = confusion_matrix(y_test, y_pred_test)
print(cm)

# その他の評価指標の表示 (Precision, Recall, F1-score)
print("\nテストデータの分類レポート:")
print(classification_report(y_test, y_pred_test))

1. データの読み込みと前処理...
訓練データ: 711件, テストデータ: 178件

2. AdaBoostモデルの学習 (scikit-learn版)...
学習完了．

3. モデルの評価...
訓練データの正解率 (Accuracy): 0.8200
テストデータの正解率 (Accuracy): 0.8258

テストデータの混同行列:
[[92 17]
 [14 55]]

テストデータの分類レポート:
              precision    recall  f1-score   support

           0       0.87      0.84      0.86       109
           1       0.76      0.80      0.78        69

    accuracy                           0.83       178
   macro avg       0.82      0.82      0.82       178
weighted avg       0.83      0.83      0.83       178



出力結果から，訓練データの正解率は82.0%，テストデータの正解率は82.6%と，両者がほぼ一致していることが確認できます．これは過学習が生じていないことを示しています．

混同行列を見ると，テストデータ178件のうち，実際に生存しなかった乗客（クラス0）109件のうち92件を正しく予測（TN），実際に生存した乗客（クラス1）69件のうち55件を正しく予測（TP）しています．

分類レポートでは，生存しなかった乗客（クラス0）に対する適合率（precision）が0.87と高く，生存した乗客（クラス1）の再現率（recall）が0.80であることから，モデルは全体的にバランスの取れた性能を示していることがわかります．